op-batcher: overhaul throttling flags

geoknee opened the pull request:
This PR reworks the throttling configuration in the batcher, starting from what makes sense at the `flags` level and then adopting the new terms throughout the codebase (hence large diff). 

It also changes the behaviour for tx size throttling, we now interpolate between two values (like block size throttling) but do not set any default limit on transaction size (unlike block size throttling). 

<img width="644" height="810" alt="throttling" src="https://github.com/user-attachments/assets/82eca948-a290-47fc-a586-ed67590a4476" />


TODO 
- [x] push all new config vars down from CLI level into business logic
- [x] ensure transaction size throttling follows new semantics

geoknee committed:
pinch throttle flags into a separate file and organize

* add new diagram which contains all 6 config vars of interest

geoknee committed:
fixes and break out default consts

geoknee committed:
WIP

geoknee committed:
WIP

geoknee committed:
fix flag env var

geoknee committed:
more fixes

geoknee committed:
WIP

geoknee committed:
use interpolation for both transactions and block size

geoknee committed:
fixes

geoknee committed:
fixes

geoknee committed:
fixes

geoknee committed:
fixes

geoknee committed:
lint

geoknee commented on op-batcher/flags/throttle_flags.go:
```diff
@@ -0,0 +1,171 @@
+	}
+	ThrottleUsafeDABytesLowerThresholdFlag = &cli.IntFlag{
+		Name:    "throttle.unsafe-da-bytes-lower-threshold",
+		Usage:   "The threshold on unsafe_da_bytes beyond which the batcher will start to throttle the block builder. Zero disables throttling.",
```
There's an opportunity here to clean up the way we disable throttling. IMO cleaner to have a "disabled" controller or other more obvious way to disable rather than a magic value here.

tobidae-cb replied:
This is a great idea, providing disabled controller type to the rpc method resets throttling and keeps it disabled. It feels like a feature that should come with a lot of warnings though

geoknee replied:
Taking that on here https://github.com/ethereum-optimism/optimism/pull/17067

geoknee committed:
throttle controller: allow lower limit = 0 as special case

geoknee committed:
more checks

geoknee committed:
lint

geoknee committed:
fix wiring

geoknee committed:
fix test

geoknee commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +69,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+
+// intensityToBlockSize converts intensity in [0,1] to block size
+func (tc *ThrottleController) intensityToBlockSize(intensity float64, cfg ThrottleConfig) uint64 {
+	if cfg.BlockSizeLowerLimit == 0 {
+		return 0
+	}
```
We should ensure this special case is well documented.

geoknee commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +69,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+
+// intensityToTxSize converts intensity in [0,1] to tx size
+func (tc *ThrottleController) intensityToTxSize(intensity float64, cfg ThrottleConfig) uint64 {
+	if cfg.TxSizeLowerLimit == 0 {
+		return 0
+	}
```
Check the documentation around this.

geoknee commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +69,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+	}
+
+	if intensity == 0 {
+		return 0 // Transactions are not throttled at 0 intensity
```
Again, this is not immediately obvious so needs documenting.

tobidae-cb commented on op-batcher/flags/throttle_flags.go:
```diff
@@ -0,0 +1,179 @@
+			return fmt.Errorf("throttle.controller-type must be one of %v, got %s", validTypes, value)
+		},
+	}
+	ThrottleUsafeDABytesLowerThresholdFlag = &cli.IntFlag{
```
nit: not sure if the `unsafe` name is a typo or on purpose

geoknee replied:
It is on purpose, since this PR https://github.com/ethereum-optimism/optimism/pull/16950 changed from using "pending da bytes" to "unsafe da bytes".

sebastianst commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +69,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+	}
+}
+
+func (tc *ThrottleController) validateConfig(cfg ThrottleConfig) error {
+	if cfg.BlockSizeLowerLimit > 0 && cfg.BlockSizeLowerLimit >= cfg.BlockSizeUpperLimit {
+		log.Error("throttler: invalid block size limits",
+			"blockSizeLowerLimit", cfg.BlockSizeLowerLimit,
+			"blockSizeUpperLimit", cfg.BlockSizeUpperLimit,
+			"controllerType", tc.GetType(),
+		)
+		return errors.New("throttler: invalid block size limits")
+	}
+
+	if cfg.TxSizeLowerLimit > 0 &&
+		tc.GetType() != config.StepControllerType &&
+		cfg.TxSizeLowerLimit >= cfg.TxSizeUpperLimit {
+		log.Error("throttler: invalid tx size limits",
+			"txSizeLowerLimit", cfg.TxSizeLowerLimit,
+			"txSizeUpperLimit", cfg.TxSizeUpperLimit,
+			"controllerType", tc.GetType(),
+		)
+		return errors.New("throttler: invalid tx size limits")
+	}
+	return nil
+}
```
Looks like there's some duplication in checks with the new throttling config `ThrottleConfig.Check()` function. Could this be unified?

geoknee replied:
I deduped the repeated bits here [87bb8ef](https://github.com/ethereum-optimism/optimism/pull/17032/commits/87bb8effba930afeef3a473627d78ed246fb2773)

sebastianst commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -50,9 +51,11 @@ func (tc *ThrottleController) Update(currentPendingBytes uint64) ThrottleParams
-	maxBlockSize := config.AlwaysBlockSize
-	var maxTxSize uint64 = 0
+func (tc *ThrottleController) intensityToParams(intensity float64, cfg ThrottleConfig) ThrottleParams {
+	err := tc.validateConfig(cfg)
```
Why do we validate on each `intensityToParams` call? Wouldn't that be every time throttling is happening? Can't we just validate once at startup?

geoknee replied:
Yes that is possible, although we also need to check at runtime because the config can be changed at runtime. Implemented that here [87bb8ef](https://github.com/ethereum-optimism/optimism/pull/17032/commits/87bb8effba930afeef3a473627d78ed246fb2773)

sebastianst replied:
Thanks, I think that's much better, so it's only checked once each time the controller changes

sebastianst commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +69,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+		return 0
+	}
+
+	if intensity == 0 {
```
what's the intensity of the step controller?

geoknee replied:
It has an intensity of 0 until the unsafe da bytes goes over the lower threshold, then it has an intensity of 1 above that.

sebastianst commented on op-batcher/batcher/throttler/controller_test.go:
```diff
@@ -456,50 +466,55 @@ func TestIntensityToParams(t *testing.T) {
 		intensity            float64
+		expectedMaxTxSize    uint64
 		expectedMaxBlockSize uint64
 		tolerance            uint64
```
why does the test need a tolerance?

geoknee replied:
I think because we are doing some floating point arithmetic and then converting back to an integer. Regardless, I don't think we need a tolerance here so I removed it [0364ab6](https://github.com/ethereum-optimism/optimism/pull/17032/commits/0364ab63819b5a8c09fc02c6d217607bb3c9642b) (the answer will be deterministic so if there is a rounding error in future test cases we can just adjust the text expectation value, we can do an exact comparison to it).

sebastianst replied:
ok thanks, it confused my why we need a tolerance in a deterministic test.

sebastianst commented on op-batcher/flags/throttle_flags.go:
```diff
@@ -0,0 +1,179 @@
+	DefaultThrottleTxSizeLowerLimit    = 150
+	DefaultThrottleTxSizeUpperLimit    = 10_000
+	DefaultThrottleBlockSizeLowerLimit = 21_000
+	DefaultThrottleBlockSizeUpperLimit = 130_000_000
```
The old limit was `130_000`, why 1000x larger now?

geoknee replied:
Good catch, fixed here [3d6f521](https://github.com/ethereum-optimism/optimism/pull/17032/commits/3d6f5217b5aa4ecec04c942b20ea8c34cbfa5601)

sebastianst commented on op-batcher/flags/throttle_flags.go:
```diff
@@ -0,0 +1,179 @@
+	// Block-builder side
+	DefaultThrottleTxSizeLowerLimit    = 150
+	DefaultThrottleTxSizeUpperLimit    = 10_000
+	DefaultThrottleBlockSizeLowerLimit = 21_000
```
I think this should be lower, this would still be 10.5kB/s with the strongest throttling intensity. Maybe 1000? Just thinking of the case where 50 rollups would max-throttle, then we'd still have 50x500kB/s=25kB/s with this setting while the max blob throughput is 96kB/s, and there would be a large backlog to clear first.

geoknee replied:
I think you meant to say 50*10KB/s = 500KB/s. If max blob throughput is about 100 kB/s then we need to reduce by a factor of 5?

Working it back the other way, if we set the lower limit to 4000 bytes, then this would imply a throughput of 2KB/s, allowing for 96/2 = 48 rollups to operate at this level simultaneously. This is roughly the throughput we see on OPM.

geoknee replied:
I went with 4000 bytes for now [3d6f521](https://github.com/ethereum-optimism/optimism/pull/17032/commits/3d6f5217b5aa4ecec04c942b20ea8c34cbfa5601)

sebastianst replied:
I meant `50x500B/s` because a block limit of 1000 with a block time of 2s is `500B/s` per rollup.

4000 works, but means max throttling just leaves enough throughput to handle max-throttled throughput. But at such a point, many rollups may have many MB each of backlog, so I thought in the extreme case, throttling could be stronger than just barely enough.

So maybe we pick the middle, 2000?

geoknee replied:
Yeah that works [edf8119](https://github.com/ethereum-optimism/optimism/pull/17032/commits/edf811969c350d14f4f2d75c88b56eb2aad83c43)

sebastianst commented on op-e2e/system/e2esys/setup.go:
```diff
@@ -576,10 +576,11 @@ func WithBatcherCompressionAlgo(ca derive.CompressionAlgo) StartOption {
-			cfg.ThrottleControllerType = batcherCfg.StepControllerType
-			cfg.ThrottleTxSize = txSize
-			cfg.ThrottleBlockSize = blockSize
+			cfg.ThrottleConfig.LowerThreshold = threshold
```
so the upper threshold can be omitted for the step controller?

geoknee replied:
Correct. The flag for the upper threshold states it is only required for linear and quadratic controllers.

sebastianst commented on op-batcher/throttling.md:
```diff
@@ -15,10 +15,22 @@ The throttling system prevents these issues by instructing sequencers to limit b
+> `throttle.block-size-lower-limit` and `throttle.tx-size-lower-limit`
+> will be interpreted by block builders as "no limit"
 
 ### Step Controller (Default)
```
Maybe we should put a warning that the step controller should not be used with default values? Because it would too soon throttle too hard?

geoknee replied:
Done

geoknee committed:
add detail to flag docs

geoknee committed:
deduplicate config validation and do not check inside intensityToParams

geoknee committed:
change defaults

block size upper limit restored to previous value of 130K

block size lower limit reduced to 4K

geoknee committed:
tweak docs

geoknee commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +64,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+		return errors.New("throttler: invalid block size limits")
+	}
+
+	if cfg.TxSizeLowerLimit > 0 &&
+		tc.GetType() != config.StepControllerType &&
+		cfg.TxSizeLowerLimit >= cfg.TxSizeUpperLimit {
```
The step contoller does not use `TxSizeUpperLimit`

https://github.com/ethereum-optimism/optimism/blob/87bb8effba930afeef3a473627d78ed246fb2773/op-batcher/batcher/throttler/controller.go#L116-L132

geoknee commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +64,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+}
+
+func (tc *ThrottleController) validateConfig(cfg ThrottleConfig) error {
+	if cfg.BlockSizeLowerLimit > 0 && cfg.BlockSizeLowerLimit >= cfg.BlockSizeUpperLimit {
```
We error if the lower limits are greater than or equal to the upper limits. It is therefore not possible for the upper limits to be zero without causing an error.

geoknee commented on op-batcher/batcher/throttler/controller.go:
```diff
@@ -66,25 +64,70 @@ func (tc *ThrottleController) intensityToParams(intensity float64, config Thrott
+		case config.StepControllerType:
+			return cfg.TxSizeLowerLimit
+		default:
+			return uint64(float64(cfg.TxSizeUpperLimit) - intensity*float64(cfg.TxSizeUpperLimit-cfg.TxSizeLowerLimit))
```
I don't see how this could underflow or overflow, unless the user tries to set TxSizeUpperLimit to a number greater than the max uint64. This would result in a parse error from the CLI library.

geoknee committed:
change flag types to uint64

geoknee committed:
remove tolerance from test

sebastianst commented on op-batcher/batcher/config.go:
```diff
@@ -53,15 +53,6 @@ func (c *ThrottleConfig) Check() error {
-	if c.TxSizeLowerLimit > 0 && c.TxSizeLowerLimit >= c.TxSizeUpperLimit {
-		return fmt.Errorf("throttle.tx-size-lower-limit must be less than throttle.tx-size-upper-limit")
-	}
-
```
I see the issue, we're dealing with different config struct in different packages...
So removing this here means, the `Check` won't check for these rules early? Maybe we then better put it back in. I thought there was a way to verify without duplicating. I guess if there was a function `throttler.VerifyParams(lowerThreshold, ...)` then both configs' verification functions could call it. But maybe that's a bit over the top. So ok to add these checks back in.

geoknee replied:
[913d8d7](https://github.com/ethereum-optimism/optimism/pull/17032/commits/913d8d79e7ab48eb71cc92281ef3b241ed2a6207)

sebastianst commented on op-batcher/throttling.md:
```diff
@@ -38,6 +38,9 @@ diagram:
 - **Use case**: Simple, predictable throttling behavior
 - **Best for**: Environments requiring clear, binary throttling states
 
+> WARNING
```
```suggestion
> [!WARNING]
```
Then we'll get a [nicely formatted alert callout](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts).

geoknee committed:
reduce block size lower limit default further to 2K

geoknee committed:
Update op-batcher/throttling.md

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

geoknee committed:
reinstate checks
